File: leetcode/top-interview-questions-easy/arrays/6. intersections_of_two_arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function (nums1, nums2) {
  var map = Object.create(null);
  var result = [];
  for (var num1 of nums1) {
    var count = map[num1] || 0;
    map[num1] = count + 1;
  }

  for (var num2 of nums2) {
    var count = map[num2];
    if (!count) {
      continue;
    }
    result.push(num2);
    map[num2] = count - 1;
  }

  return result;
};

File: leetcode/top-interview-questions-easy/arrays/6. intersections_of_two_arrays.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        map = {}
        result = []
        
        for num1 in nums1:
            count = map.get(num1, 0)
            map[num1] = count + 1
        
        for num2 in nums2:
            count = map.get(num2, 0)
            if not count:
                continue
            result.append(num2)
            map[num2] = count - 1
            
        return result
